File: scala/PalindromeRecursive.scala
object PalindromeRecursive {
    def isPalindrome(str: String): Boolean = {
        if (str.length == 0 || str.length == 1) { 
            return true 
        }
        if (str(0) == str(str.length-1)) {
            isPalindrome(str.substring(1, str.length-1))
        } else { false }
    }
    
    def main(args: Array[String]) = {
        println(isPalindrome("hi"))
        print(isPalindrome("ababa"))
    }
}

File: java/PalindromeRecursive.java
final class PalindromeRecursive {
    
    static boolean isPalindrome(String str) {
        if (str.length() == 1 || str.length() == 0) {
            return true;
        }
        if (str.charAt(0) == str.charAt(str.length()-1)) {
            return isPalindrome(str.substring(1,str.length()-1));
        } else {
            return false;
        }
    }

    public static void main(String[] args) {
        System.out.println(isPalindrome("hi"));
        System.out.println(isPalindrome("ababa"));
    }
}